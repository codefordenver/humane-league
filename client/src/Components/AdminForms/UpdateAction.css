.UpdateAction {
  display: flex;
  flex-direction: column;
  align-items: center; }
  .UpdateAction .update-container {
    margin: 20px;
    padding: 20px;
    position: relative;
    background-color: #ffffffe3;
    border-radius: 2px;
    box-shadow: 2px 2px 15px rgba(0, 0, 0, 0.6); }
    .UpdateAction .update-container .update-controls {
      display: flex; }
      .UpdateAction .update-container .update-controls label {
        font-size: 20px;
        margin: 0 15px; }
      .UpdateAction .update-container .update-controls select,
      .UpdateAction .update-container .update-controls option {
        cursor: pointer;
        font-size: 20px;
        height: 40px;
        padding: 5px;
        margin: 20px; }
    .UpdateAction .update-container .sort-by {
      width: 50%;
      display: flex;
      flex-direction: column; }
      .UpdateAction .update-container .sort-by p {
        font-size: 20px; }
      .UpdateAction .update-container .sort-by label,
      .UpdateAction .update-container .sort-by input {
        margin: 0 15px; }
    .UpdateAction .update-container .edit-action-form {
      position: absolute;
      top: -35px;
      left: -75px;
      margin: 20px;
      padding: 40px 20px;
      background-color: #ffffffe3;
      border-radius: 2px;
      box-shadow: 2px 2px 15px rgba(0, 0, 0, 0.6); }
      .UpdateAction .update-container .edit-action-form h3 {
        text-align: center;
        background: #aaa;
        padding: 10px; }
      .UpdateAction .update-container .edit-action-form button.cancel {
        margin: 0 auto;
        font-size: 14pt;
        padding: 10px;
        color: white;
        background: #666;
        cursor: pointer;
        text-decoration: none;
        width: 94%;
        display: flex;
        align-items: center;
        justify-content: center;
        transition: 0.3s; }
        .UpdateAction .update-container .edit-action-form button.cancel:hover {
          color: #000;
          background: #aaa; }
    .UpdateAction .update-container .actions-container {
      margin: 20px; }
      .UpdateAction .update-container .actions-container h3 {
        text-align: center; }
      .UpdateAction .update-container .actions-container .actions {
        list-style: none; }
        .UpdateAction .update-container .actions-container .actions .action {
          cursor: pointer;
          background: #fff;
          border: 1px solid darkgray;
          font-size: 20px;
          margin: 10px; }
          .UpdateAction .update-container .actions-container .actions .action p {
            padding: 10px; }
          .UpdateAction .update-container .actions-container .actions .action:hover {
            background: #009639;
            box-shadow: 0px 0px 5px rgba(0, 0, 0, 0.6); }
          .UpdateAction .update-container .actions-container .actions .action:hover p {
            color: #fff;
            font-weight: bold; }
        .UpdateAction .update-container .actions-container .actions .action.disabled {
          background: #bbb; }
          .UpdateAction .update-container .actions-container .actions .action.disabled:hover {
            background: #999; }
