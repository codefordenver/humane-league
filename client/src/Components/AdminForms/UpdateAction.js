import React, { Component } from 'react';
import { connect } from 'react-redux';

// import './UpdateAction.css';

export class UpdateAction extends Component {
  constructor() {
    super();
    this.state = {
      actionType: 'facebook',
      action: {},
      showForm: false,
      facebook: [],
      twitter: [],
      email: [],
      phone: []
    }
  }

  async componentDidMount() {
    const facebookFetch = await fetch('/api/v1/facebook_actions');
    const facebookActions = await facebookFetch.json();
    const facebook = facebookActions.results;

    const twitterFetch = await fetch('/api/v1/twitter_actions');
    const twitterActions = await twitterFetch.json();
    const twitter = twitterActions.results;

    const emailFetch = await fetch('/api/v1/email_actions');
    const emailActions = await emailFetch.json();
    const email = emailActions.results;

    const phoneFetch = await fetch('/api/v1/phone_actions');
    const phoneActions = await phoneFetch.json();
    const phone = phoneActions.results;
    console.log(phone);
    await this.setState({ facebook, twitter, email, phone });
  };

  handleChange = () => {
    const actionType = this.updateActionTypes.value;
    
    this.setState({ actionType });
  }

  handleActionClick = (event) => {
    const actionId = event.target.dataset.id;
    const action = this.state[this.state.actionType].find(action => action.id == actionId);

    this.setState({ showForm: true, actionEnabled: action.enabled, action });
  }

  submitPatch = async () => {
    const oldAction = this.state.action;
    const type = this.state.actionType;

    const action = Object.assign({ ...oldAction }, { enabled: this.state.actionEnabled });
    const actionId = action.id;

    const token = this.props.user.id_token;
    const actionPatch = await fetch(`/api/v1/${type}_actions/${actionId}?token=${token}`, {
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(action)
    });

    if (actionPatch.status === 204) {
      const removedAction = this.state[this.state.actionType].filter(action => action.id !== actionId);
      const newState = [...removedAction, action];
      this.setState({ [type]: newState, showForm: false, success: `Sucessfully updated action ${action.id}`});
        setTimeout(() => {
        this.setState({ success: false });
      }, 3000);
    }
  }

  render() {
    const actions = this.state[this.state.actionType].map((action, i) => {
      return (
        <li key={`li-${i}`} className='action'>
          <p data-id={action.id} onClick={this.handleActionClick}>{`ACTION ${i}: ${action.title}`}</p>
        </li>
      )
    });

    return (
      <div className='UpdateAction'>
        <h1>Enable and Disable <span>{this.state.actionType.toUpperCase()}</span> Actions</h1>
        <label htmlFor='action-types'>Select Action Type:
          <select onChange={this.handleChange} ref={(elem) => {this.updateActionTypes = elem}} name='action-types' id='action-types'>
            <option value='facebook'>Facebook</option>
            <option value='twitter'>Twitter</option>
            <option value='email'>Email</option>
            <option value='phone'>Phone</option>
          </select>
        </label>
        {
          this.state.success && 
          <p>{this.state.success}</p>
        }
        {
          this.state.showForm &&
          <div className='update-form'>   
            <span id='toggle'>
              <input onChange={() => this.setState({ actionEnabled: !this.state.actionEnabled })} checked={this.state.actionEnabled} ref={(elem) => {this.toggle = elem}} type='checkbox'/>
              <label 
                data-on='enabled' 
                data-off='disabled'>
              </label>
            </span>
            <button onClick={this.submitPatch} className='update-action'>Save Update</button>
          </div>
        }
        <div className='actions-container'>
          <ul className='actions'>
            {actions}
          </ul>
        </div>
      </div>
    )
  }
}

const mapStateToProps = store => ({
  user: store.User
});

export default connect(mapStateToProps, null)(UpdateAction);