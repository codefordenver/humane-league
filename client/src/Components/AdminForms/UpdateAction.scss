.UpdateAction {
  display: flex;
  flex-direction: column;
  align-items: center;
  .update-container {
    margin: 20px;
    padding: 20px;
    background-color: #ffffffe3;
    border-radius: 2px;
    box-shadow: 2px 2px 15px rgba(0, 0, 0, 0.6);
    .update-controls {
      display: flex;

      label {
        font-size: 20px;
        margin: 0 15px;
      }
      select,
      option {
        cursor: pointer;
        font-size: 20px;
        height: 40px;
        padding: 5px;
        margin: 20px;
      }
    }
    .sort-by {
      width: 50%;
      display: flex;
      flex-direction: column;
      p {
        font-size: 20px;
      }
      label,
      input {
        margin: 0 15px;
      }
    }
    .actions-container {
      margin: 20px;
      h3 {
        text-align: center;
      }
      .actions {
        list-style: none;
        .action {
          cursor: pointer;
          background: #fff;
          border: 1px solid darkgray;
          font-size: 20px;
          margin: 10px;
          p {
            padding: 10px;
          }
          &:hover{
            background: #009639;
            box-shadow: 0px 0px 5px rgba(0, 0, 0, 0.6);
          }
          &:hover p {
            color: #fff;
            font-weight: bold;
          }
        }
        .action.disabled {
          background: #bbb;
          &:hover {
            background: #999;
          }
        }
      }
    }
  }
}