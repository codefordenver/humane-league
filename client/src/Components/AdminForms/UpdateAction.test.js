import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import { UpdateAction } from './UpdateAction';

describe.skip('UpdateAction component tests', () => {
  let renderedUpdateAction;
  let defaultState; 
  
  beforeEach(() => {
    renderedUpdateAction = shallow( <UpdateAction user={{ name: 'hello' }}/> );
    defaultState = {
      actionType: 'facebook',
      action: {},
      showForm: false,
      facebook: [],
      twitter: [],
      email: [],
      phone: []
    }

    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => 
        Promise.resolve(
          [{id: 1, title: 'title'}]
        )   
    }));
  })
  it('renders without crashing', () => {
    expect(renderedUpdateAction).toBeDefined();
    expect(renderedUpdateAction.state()).toEqual(defaultState);
  });
  it('matches SnapShot', () => {
    // expect(renderedUpdateAction).toMatchSnapshot();
  })
})