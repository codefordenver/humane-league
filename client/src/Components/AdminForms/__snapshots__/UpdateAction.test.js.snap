// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateAction component tests matches SnapShot 1`] = `undefined`;
